src/Microsoft.ML.Core/Utilities/LineParser.cs

```diff
@@ -0,0 +1,55 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.ML.Runtime.Internal.Utilities
+{
+    [BestFriend]
+    internal static class LineParser
+    {
+        [BestFriend] // required because the return type is a value tuple..
+        internal static (bool isSuccess, string key, float[] values) ParseKeyThenNumbers(string line)
+        {
+            if (string.IsNullOrWhiteSpace(line))
+                return (false, null, null);
+
+            ReadOnlySpan<char> trimmedLine = line.AsSpan().TrimEnd(); // TrimEnd creates a Span, no allocations
+
+            int firstSeparatorIndex = trimmedLine.IndexOfAny(' ', '\t'); // the first word is the key, we just skip it
+            ReadOnlySpan<char> valuesToParse = trimmedLine.Slice(start: firstSeparatorIndex + 1);
+
+            float[] values = AllocateFixedSizeArrayToStoreParsedValues(valuesToParse);
+
+            int toParseStartIndex = 0;
+            int valueIndex = 0;
+            for (int i = 0; i <= valuesToParse.Length; i++)
+            {
+                if (i == valuesToParse.Length || valuesToParse[i] == ' ' || valuesToParse[i] == '\t')
+                {
+                    if (DoubleParser.TryParse(valuesToParse.Slice(toParseStartIndex, i - toParseStartIndex), out float parsed))
+                        values[valueIndex++] = parsed;
+                    else
+                        return (false, null, null);
+
+                    toParseStartIndex = i + 1;
+                }
+            }
+
+            return (true, trimmedLine.Slice(0, firstSeparatorIndex).ToString(), values);
```

Comment:
I'm not sure how important this is (but since we are doing this in the name of perf):

The `string key` allocates a new string here, and later is used to add to a `NormStr.Pool`, which can operate over `ReadOnlyMemory<char>`.

If we returned a `ROM<char>` here instead, we wouldn't need to allocate a new string at all.

Reply:
I did not show in the profiles. I will give it a try in round 2 of the optimizations.